gobblin-data-management/src/main/java/gobblin/data/management/copy/replication/ConfigBasedDatasetsFinder.java

```diff
@@ -0,0 +1,181 @@
+  public static final String CONFIG_STORE_REPLICATION_TAG = "config.store.replication.tag";
+  public static final String CONFIG_STORE_REPLICATION_DISABLE_TAGS = "config.store.replication.disable.tags";
+
+  private final String storeRoot;
+  private final Path replicationRootPath;
+  private final Path replicationTag;
+  private final Optional<List<Path>> replicationDisableTags;
+  private final ConfigClient configClient;
+
+  public ConfigBasedDatasetsFinder(Properties props) throws IOException {
+    Preconditions.checkArgument(props.containsKey(CONFIG_STORE_ROOT),
+        "missing required config entery " + CONFIG_STORE_ROOT);
+    Preconditions.checkArgument(props.containsKey(CONFIG_STORE_REPLICATION_ROOT),
+        "missing required config entery " + CONFIG_STORE_REPLICATION_ROOT);
+    Preconditions.checkArgument(props.containsKey(CONFIG_STORE_REPLICATION_TAG),
+        "missing required config entery " + CONFIG_STORE_REPLICATION_TAG);
+
+    this.storeRoot = props.getProperty(CONFIG_STORE_ROOT);
+    this.replicationRootPath =
+        PathUtils.mergePaths(new Path(this.storeRoot), new Path(props.getProperty(CONFIG_STORE_REPLICATION_ROOT)));
+    this.replicationTag =
+        PathUtils.mergePaths(new Path(this.storeRoot), new Path(props.getProperty(CONFIG_STORE_REPLICATION_TAG)));
+    
+    if(props.containsKey(CONFIG_STORE_REPLICATION_DISABLE_TAGS)){
+      List<String> disableStrs = Splitter.on(",").omitEmptyStrings().splitToList(props.getProperty(CONFIG_STORE_REPLICATION_DISABLE_TAGS));
+      List<Path> disablePaths = new ArrayList<Path>();
+      for(String s: disableStrs){
+        disablePaths.add(PathUtils.mergePaths(new Path(this.storeRoot), new Path(s)));
+      }
+      this.replicationDisableTags = Optional.of(disablePaths);
+    }
+    else{
+      this.replicationDisableTags = Optional.absent();
+    }
+
+    configClient = ConfigClient.createConfigClient(VersionStabilityPolicy.WEAK_LOCAL_STABILITY);
+  }
+
+  protected static Set<URI> getValidDatasetURIs(Collection<URI> allDatasetURIs, Collection<URI> disabledURIs, Path ancestor) {
+    if (allDatasetURIs == null || allDatasetURIs.isEmpty()) {
+      return ImmutableSet.of();
+    }
+
+    Comparator<URI> pathLengthComparator = new Comparator<URI>() {
+      public int compare(URI c1, URI c2) {
+        return c1.getPath().length() - c2.getPath().length();
+      }
+    };
+
+    List<URI> datasetsList = new ArrayList<URI>(allDatasetURIs);
+
+    // sort the URI based on the path length to make sure the parent path appear before children
+    Collections.sort(datasetsList, pathLengthComparator);
+
+    Set<URI> disabledSet = new HashSet<URI>(disabledURIs);
+    Set<URI> validURISet = new HashSet<URI>();
+
+    // remove non leaf URI
+    for (URI u : datasetsList) {
+      URI needToRemove = null;
+
+      // valid dataset URI
+      if (PathUtils.isAncestor(ancestor, new Path(u.getPath()))) {
+        
+        for (URI leaf : validURISet) {
```

Comment:
Current algorithm is a O(n2). You can make it O(nlog n) by a TreeSet for validURISet and exploiting the fact that the necessary condition for a path PF to be a prefix of a path P is to be the largest path smaller that P, i.e.  PF=floor(P).